File: src/ai-service/telemetry.py
"""
OpenTelemetry auto-instrumentation wrapper for ai-service
"""

import os
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
# Using aiohttp-client-instrumentation package
# Note: aiohttp instrumentation requires opentelemetry-instrumentation-aiohttp-client package
try:
    from opentelemetry.instrumentation.aiohttp_client import AioHttpClientInstrumentor
    AIOHTTP_AVAILABLE = True
except ImportError:
    AIOHTTP_AVAILABLE = False
from opentelemetry.semconv.resource import ResourceAttributes


def configure_telemetry(app):
    """Configure OpenTelemetry auto-instrumentation for the FastAPI application"""
    try:
        # Get collector endpoint from environment or use default
        otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "localhost:4317")
        
        # Get service details from environment
        service_name = "ai-service"
        service_version = os.getenv("APP_VERSION", "0.1.0")
        environment = os.getenv("ENVIRONMENT", "development")
        
        # Configure resource attributes
        resource = Resource.create({
            ResourceAttributes.SERVICE_NAME: service_name,
            ResourceAttributes.SERVICE_VERSION: service_version,
            "environment": environment,
        })
        
        # Create a TracerProvider with the resource
        tracer_provider = TracerProvider(resource=resource)
        
        # Create and add span processor with OTLP exporter
        span_processor = BatchSpanProcessor(
            OTLPSpanExporter(endpoint=otlp_endpoint, insecure=True)
        )
        tracer_provider.add_span_processor(span_processor)
        
        # Set the tracer provider
        trace.set_tracer_provider(tracer_provider)
        
        # Instrument FastAPI
        FastAPIInstrumentor.instrument_app(app)
        
        # Instrument other libraries
        RequestsInstrumentor().instrument()
        if AIOHTTP_AVAILABLE:
            AioHttpClientInstrumentor().instrument()
        
        print("OpenTelemetry instrumentation initialized for ai-service")
        return tracer_provider
    except Exception as e:
        print(f"Failed to initialize OpenTelemetry instrumentation: {e}")
        return None

def shutdown_telemetry(tracer_provider):
    """Shutdown OpenTelemetry tracer provider gracefully"""
    if tracer_provider:
        try:
            tracer_provider.shutdown()
            print("OpenTelemetry tracer provider shut down successfully")
        except Exception as e:
            print(f"Error shutting down tracer provider: {e}")

File: src/store-front/src/telemetry.ts
import { WebTracerProvider } from '@opentelemetry/sdk-trace-web';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { ZoneContextManager } from '@opentelemetry/context-zone';
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { getWebAutoInstrumentations } from '@opentelemetry/auto-instrumentations-web';

// Initialize the OpenTelemetry provider
export function initTelemetry() {
  console.log('OpenTelemetry initialization started for store-front');
  
  try {
    // Configure resource
    const resource = new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: 'store-front',
      [SemanticResourceAttributes.SERVICE_VERSION]: '0.1.0',
      'environment': 'development'
    });

    // Create tracer provider
    const provider = new WebTracerProvider({ resource });

    // Configure OTLP exporter
    const exporter = new OTLPTraceExporter({
      url: 'http://localhost:4318/v1/traces',
    });

    // Add span processor
    provider.addSpanProcessor(new BatchSpanProcessor(exporter));

    // Register the provider
    provider.register({
      contextManager: new ZoneContextManager(),
    });

    // Register auto-instrumentations
    registerInstrumentations({
      instrumentations: [getWebAutoInstrumentations()],
    });

    const tracer = provider.getTracer('store-front');
    
    console.log('OpenTelemetry successfully initialized for store-front');
    
    return { 
      tracer, 
      provider 
    };
  } catch (error) {
    console.error('Failed to initialize OpenTelemetry:', error);
    console.log('Application will continue without OpenTelemetry instrumentation');
    
    return { 
      tracer: undefined, 
      provider: undefined 
    };
  }
}

// Cleanup function to be called when the application is unmounting
export function shutdownTelemetry(provider?: WebTracerProvider) {
  if (provider) {
    try {
      provider.shutdown()
        .then(() => console.log('OpenTelemetry provider shut down successfully'))
        .catch(error => console.error('Error shutting down OpenTelemetry provider:', error));
    } catch (error) {
      console.error('Error during OpenTelemetry shutdown:', error);
    }
  }
}
